Unify TensorOptions signatures

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40714 Change to.dtype_layout to c10-full
* **#39611 Unify TensorOptions signatures**

A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

This should be backwards compatible because non-optional things are a subtype of optional things in the binary representation on the stack, i.e. if you read an optional from the stack but there was a non-optional, that's fine, so unboxing is BC. Boxing is not BC because it writes to the stack and that's contravariant, but for loading mobile models only the BC of unboxing is important.

This PR will break forward compatibility but only in the sense that newly written models might use the new feature of calling the ops with optional values and those new models will not be runnable on old PyTorch versions. Any model existing today will not be FC broken and can be serialized with a new PyTorch to run on older PyTorch versions.

Detailed reasoning:
---
This doesn't break BC because an IValue holding a optional<ScalarType> that is defined looks exactly the same as an IValue holding a non-optional ScalarType. So any old model that got serialized and expects to call the non-optional operators will be able to call the optional operators and PyTorch won't be able to tell the difference.

This also doesn't immediately break FC since any model written today will only call those ops with defined values, so even if a pre-existing model gets re-serialized after my change, the values will be serialized as defined values and an old pytorch will be able to call the non-optional version of the operator with it.

However, after this change, users will be able to write new models that call these operators with None values. And if somebody writes such a model, then it will not be loadable on older PyTorch. So in a sense, it is FC breaking.


Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

smessmer committed:
Unify TensorOptions signatures

A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

ezyang commented:
The last time we attempted this was https://github.com/pytorch/pytorch/pull/34250  I think there is a reason you cannot conveniently do this. @bhosmer might remember more clearly.

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3598,6 +3602,8 @@
 # See NOTE [ TensorOptions Constructors ].
 - func: to.dtype_layout(Tensor self, *, ScalarType dtype, Layout layout, Device device, bool pin_memory=False, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor
+  #- func: to.dtype_layout(Tensor self, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor
+  #use_c10_dispatcher: full
```
Would be good to mention why this didn't work

smessmer replied:
I'll just remove those changes from the PR before landing and address them in a separate PR.

ezyang commented:
NB: this breaks forward compatibility

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

This should be backwards compatible because non-optional things are a subtype of optional things in the binary representation on the stack, i.e. if you read an optional from the stack but there was a non-optional, that's fine, so unboxing is BC. Boxing is not BC because it writes to the stack and that's contravariant, but for loading mobile models only the BC of unboxing is important.

This PR will break forward compatibility.

Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

This should be backwards compatible because non-optional things are a subtype of optional things in the binary representation on the stack, i.e. if you read an optional from the stack but there was a non-optional, that's fine, so unboxing is BC. Boxing is not BC because it writes to the stack and that's contravariant, but for loading mobile models only the BC of unboxing is important.

This PR will break forward compatibility but only in the sense that newly written models might use the new feature of calling the ops with optional values and those new models will not be runnable on old PyTorch versions. Any model existing today will not be FC broken and can be serialized with a new PyTorch to run on older PyTorch versions.

Detailed reasoning:
---
This doesn't break BC because an IValue holding a optional<ScalarType> that is defined looks exactly the same as an IValue holding a non-optional ScalarType. So any old model that got serialized and expects to call the non-optional operators will be able to call the optional operators and PyTorch won't be able to tell the difference.

This also doesn't immediately break FC since any model written today will only call those ops with defined values, so even if a pre-existing model gets re-serialized after my change, the values will be serialized as defined values and an old pytorch will be able to call the non-optional version of the operator with it.

However, after this change, users will be able to write new models that call these operators with None values. And if somebody writes such a model, then it will not be loadable on older PyTorch. So in a sense, it is FC breaking.


Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

This should be backwards compatible because non-optional things are a subtype of optional things in the binary representation on the stack, i.e. if you read an optional from the stack but there was a non-optional, that's fine, so unboxing is BC. Boxing is not BC because it writes to the stack and that's contravariant, but for loading mobile models only the BC of unboxing is important.

This PR will break forward compatibility but only in the sense that newly written models might use the new feature of calling the ops with optional values and those new models will not be runnable on old PyTorch versions. Any model existing today will not be FC broken and can be serialized with a new PyTorch to run on older PyTorch versions.

Detailed reasoning:
---
This doesn't break BC because an IValue holding a optional<ScalarType> that is defined looks exactly the same as an IValue holding a non-optional ScalarType. So any old model that got serialized and expects to call the non-optional operators will be able to call the optional operators and PyTorch won't be able to tell the difference.

This also doesn't immediately break FC since any model written today will only call those ops with defined values, so even if a pre-existing model gets re-serialized after my change, the values will be serialized as defined values and an old pytorch will be able to call the non-optional version of the operator with it.

However, after this change, users will be able to write new models that call these operators with None values. And if somebody writes such a model, then it will not be loadable on older PyTorch. So in a sense, it is FC breaking.


Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

This should be backwards compatible because non-optional things are a subtype of optional things in the binary representation on the stack, i.e. if you read an optional from the stack but there was a non-optional, that's fine, so unboxing is BC. Boxing is not BC because it writes to the stack and that's contravariant, but for loading mobile models only the BC of unboxing is important.

This PR will break forward compatibility but only in the sense that newly written models might use the new feature of calling the ops with optional values and those new models will not be runnable on old PyTorch versions. Any model existing today will not be FC broken and can be serialized with a new PyTorch to run on older PyTorch versions.

Detailed reasoning:
---
This doesn't break BC because an IValue holding a optional<ScalarType> that is defined looks exactly the same as an IValue holding a non-optional ScalarType. So any old model that got serialized and expects to call the non-optional operators will be able to call the optional operators and PyTorch won't be able to tell the difference.

This also doesn't immediately break FC since any model written today will only call those ops with defined values, so even if a pre-existing model gets re-serialized after my change, the values will be serialized as defined values and an old pytorch will be able to call the non-optional version of the operator with it.

However, after this change, users will be able to write new models that call these operators with None values. And if somebody writes such a model, then it will not be loadable on older PyTorch. So in a sense, it is FC breaking.


Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify TensorOptions signatures"


A few ops have been taking non-optional ScalarType, Device and Layout. That isn't supported by the hacky wrapper that makes those
kernels work with the c10 operator library. This PR unifies the signatures and makes those ops c10-full.

This should be backwards compatible because non-optional things are a subtype of optional things in the binary representation on the stack, i.e. if you read an optional from the stack but there was a non-optional, that's fine, so unboxing is BC. Boxing is not BC because it writes to the stack and that's contravariant, but for loading mobile models only the BC of unboxing is important.

This PR will break forward compatibility but only in the sense that newly written models might use the new feature of calling the ops with optional values and those new models will not be runnable on old PyTorch versions. Any model existing today will not be FC broken and can be serialized with a new PyTorch to run on older PyTorch versions.

Detailed reasoning:
---
This doesn't break BC because an IValue holding a optional<ScalarType> that is defined looks exactly the same as an IValue holding a non-optional ScalarType. So any old model that got serialized and expects to call the non-optional operators will be able to call the optional operators and PyTorch won't be able to tell the difference.

This also doesn't immediately break FC since any model written today will only call those ops with defined values, so even if a pre-existing model gets re-serialized after my change, the values will be serialized as defined values and an old pytorch will be able to call the non-optional version of the operator with it.

However, after this change, users will be able to write new models that call these operators with None values. And if somebody writes such a model, then it will not be loadable on older PyTorch. So in a sense, it is FC breaking.


Differential Revision: [D21915788](https://our.internmc.facebook.com/intern/diff/D21915788/)

[ghstack-poisoned]
